tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java

```diff
@@ -0,0 +1,520 @@
+				} catch (InvalidKeyException e) {
+					throw new IOException(e);
+				} catch (NoSuchAlgorithmException e) {
+					throw new IOException(e);
+				} catch (NoSuchPaddingException e) {
+					throw new IOException(e);
+				} catch (SAXException e) {
+					throw new IOException(e);
+				} finally {
+					try {
+						input.close();
+					} catch (IOException e) {
+						log.error("never happen?", e);
+					}
+				}
+			} else {
+				log.warn("unknown Entry '{}'({})", entry.getName(), entry);
+			}
+		}
+	}
+
+	private Key readKey(InputStream input) throws IOException {
+		byte[] data = new byte[260];
+
+		input.read(data, 0, 4); // TAG,
+
+		input.read(data, 0, 256);
+
+		SRand srand = new SRand(LittleEndian.getInt(data));
+		byte xor = 0;
+		for (int i = 0, n = 0; i < 256; i++, n--) {
+			if (n == 0) {
+				xor = (byte) (srand.rand() & 0xFF);
+				n = (int) ((srand.rand() & 0xF) + 1);
+			}
+			if (i >= 4) {
+				data[i] = (byte) ((data[i]) ^ (xor));
+			}
+		}
+
+		int offset = 4 + (data[0] & 0xF);
+		byte[] key = Arrays.copyOfRange(data, offset, offset + 16);
+
+		SecretKeySpec secretKey = new SecretKeySpec(key, "AES");
+		return secretKey;
+	}
+
+	public InputStream createDecryptStream(InputStream input, Key key)
+			throws IOException, NoSuchAlgorithmException,
+			NoSuchPaddingException, InvalidKeyException {
+		Cipher cipher = null;
+
+		cipher = Cipher.getInstance("AES/ECB/NoPadding");
+		cipher.init(Cipher.DECRYPT_MODE, key);
+
+		return new CipherInputStream(input, cipher);
+	}
+
+	/**
+	 * extract characters from Section stream
+	 * 
+	 * @param reader
+	 * @param writer
+	 * @throws IOException
+	 * @throws SAXException 
+	 */
+	private void parse(HwpStreamReader reader, XHTMLContentHandler xhtml)
+			throws IOException, SAXException {
+		StringBuffer buf = new StringBuffer(1024);
+		TagInfo tag = new TagInfo();
+
+		while (true) {
+			if (!readTag(reader, tag))
+				break;
+
+			buf.setLength(0);
+			if (HWPTAG_BEGIN + 50 == tag.id) {
+				writeParaHeader(reader, tag.length, buf);
+			} else if (HWPTAG_BEGIN + 51 == tag.id) {
+				if (tag.length % 2 != 0)
+					throw new IOException("Invalid block size");
+
+				writeParaText(reader, tag.length, buf);
+
+				if (buf.length() > 0) {
+					buf.append('\n');
+					xhtml.characters(buf.toString());		
+				}
+			} else {
+				reader.ensureSkip(tag.length);
+			}
+
+			if (buf.length() > 0) {
+				log.debug("TAG[{}]({}):{} [{}]", new Object[] { tag.id,
+						tag.level, tag.length, buf });
+			}
+		}
+	}
+
+	private void writeParaHeader(HwpStreamReader reader, long length,
+			StringBuffer buf) throws IOException {
+		reader.ensureSkip(length);
+	}
+
+	/**
+	 * transfer character stream of HWPTAG_PARA_TEXT to STRING
+	 * 
+	 * @param reader
+	 * @param datasize
+	 * @param buf
+	 * @throws IOException
+	 */
+	private void writeParaText(HwpStreamReader reader, long datasize,
+			StringBuffer buf) throws IOException {
+		int[] chars = reader.uint16((int) (datasize / 2));

```

Comment:
Somewhere do a sanity/range check on datasize...apologies if you've done that earlier!

Reply:
The reader.uint16() has a sanity/range check.